src/map/skill.c

```diff
@@ -24724,14 +24724,224 @@ static bool skill_read_skilldb(const char *filename)
 		idb_iput(loaded_ids_db, tmp_db.nameid, true);
 		count++;
 	}
-	
+
 	db_destroy(loaded_ids_db);
 
 	libconfig->destroy(&skilldb);
 	ShowStatus("Done reading '"CL_WHITE"%d"CL_RESET"' entries in '"CL_WHITE"%s"CL_RESET"'.\n", count, filepath);
 	return true;
 }
 
+/**
+ * Reads a AutoSpell skill's Id (SkillId) when reading the autospell DB.
+ *
+ * @param conf The libconfig settings block which contains the skill's data.
+ * @param sk The autospell_skill struct where the id should be set.
+ */
+static void skill_read_autospell_skill_id(struct config_setting_t *conf, struct s_autospell_db *sk, int index)
+{
+	nullpo_retv(conf);
+	nullpo_retv(sk);
+
+	sk->skill_id = 0;
+
+	const char *skill_name = NULL;
+	int skill_id = 0;
+
+	if (libconfig->setting_lookup_int(conf, "SkillId", &skill_id) == CONFIG_FALSE) {
+		if (libconfig->setting_lookup_string(conf, "SkillId", &skill_name) == CONFIG_FALSE) {
+			ShowError("%s: Invalid AutoSpell db entry \"%d\". SkillId is required. Skipping...\n", __func__, index);
+			return;
+		}
+
+		skill_id = skill->name2id(skill_name);
+	}
+
+	if (skill_id == 0) {
+		if (skill_name != NULL)
+			ShowError("%s: Invalid AutoSpell db entry \"%d\". SkillId \"%s\" doesn't exists. Skipping...\n", __func__, index, skill_name);
+		else
+			ShowError("%s: Invalid AutoSpell db entry \"%d\". SkillId \"%d\" doesn't exists. Skipping...\n", __func__, index, skill_id);
+		return;
+	}
+
+	sk->skill_id = skill_id;
+}
+
+/**
+ * Reads a AutoSpell skill's usable levels (SkillLevel) when reading the autospell DB.
+ *
+ * @param conf The libconfig settings block which contains the skill's data.
+ * @param sk The autospell_skill struct where the level should be set.
+ */
+static void skill_read_autospell_skill_level(struct config_setting_t *conf, struct s_autospell_db *sk)
```

Comment:
**thought/possible improvement:** we have several functions very similar to this one in `skill.c`. I am wondering whether we could make some generalization to clear those replications...

I can see the advantage of having them separate since plugins can hook into one of the fields without worrying about the other ones.

But since we are at startup, where performance is not extremely critical, maybe we could use a parameter so plugins could simply use an if to tell whether it is the one they want. and combine this with a function received by parameter for validation/actually setting the value.

Something like: (pseudo code)
```
enum config_identifier {
   autospell_skillevel = 0,
   skill_range = 1,
   // ...
}

load_autospell_skill_level(int *arr, int level, int val, struct info) {
   // validate ... ("info" can be used to provide context for the message)
   arr[level] = val;
}

read_grouped_value(enum config_identifier, conf, field_name, load_fn) {
   // check if it is a single value or a group, read and call load_fn for each level
}
```

Reply:
I like this idea

Reply:
Thanks :) I will work on this in a separate PR so we don't mix unrelated things into the autospell one